src/merkle_tree.rs

```diff
@@ -0,0 +1,291 @@
+use std::vec;
+use hex;
+use sha2::{Digest, Sha256};
+use crate::merkle_error::MerkleError;
+use crate::proof_element::ProofElement;
+
+#[derive(Debug, PartialEq)]
+pub struct MerkleTree{
+    tree: Vec<Vec<String>>
+}
+
+impl MerkleTree{
+    /// Builds merkle tree from elements list, hashing them first.
+    pub fn build(elements: Vec<String>) -> Result<Self, MerkleError> {
+        let hashed_elements: Vec<String> = elements.iter().map(|e| Self::hash(&e)).collect();
+
+        Self::build_without_hashing(hashed_elements)
+    }
+
+    pub fn build_without_hashing(hashes: Vec<String>) -> Result<Self, MerkleError>{
+        // 1. New merkle tree
+        let mut merkle_tree = MerkleTree { tree: vec![] };
+
+        // 2. Push leaf nodes to the tree
+        let mut elements_to_push = hashes;
+
+        Self::duplicate_last_if_odd(&mut elements_to_push)?;
+        
+        merkle_tree.tree.push(elements_to_push.clone());
+        
+        // 3. Calculate upper levels and push them to the tree until root is reached.
+        while elements_to_push.len() > 1 {
+            elements_to_push = Self::calculate_upper_level(&elements_to_push);
+
+            // This "if" is because the only odd tree level accepted is the root!
+            if elements_to_push.len() > 1 { 
+                Self::duplicate_last_if_odd(&mut elements_to_push)?;
+            }
+            
+            merkle_tree.tree.push(elements_to_push.clone());
+        }
+
+        Ok(merkle_tree)
+    }
+
+    pub fn new_empty() -> MerkleTree{
+        MerkleTree { tree: vec![vec![]] }
+    }
+
+    
+    pub fn verify(&self, hash: String, proof: Vec<ProofElement>) -> Result<bool, MerkleError>{
+        // Calculates root with element hash (leaf node) and it's proof
+        let calc_root = proof.iter().fold(hash, |cur_hash, partner| { 
+            let combined_hashes = if partner.left {
+                partner.hash.clone() + &cur_hash
+            } else {
+                cur_hash + &partner.hash
+            };
+
+            MerkleTree::hash(&combined_hashes)
+        });
+
+        let real_root = self.get_root()?;
+
+        Ok(calc_root == *real_root)
+    }
+    
+    pub fn gen_proof(&self, hash: String) -> Result<Vec<ProofElement>, MerkleError> {
```

Comment:
I got kind of lost while checking the `gen_proof`, and I think it was because right now what you are doing is to rewrite `i` with `idx`, where `idx` is the partners index not the one from the hash you are actually looking to generate a proof for, what I would do is the following:

```suggestion
    pub fn gen_proof(&self, hash: String) -> Result<Vec<ProofElement>, MerkleError> {
        let mut proof: Vec<ProofElement> = vec![];
        
        let mut i = self.find_hash_index(hash)?;
        
        let mut level = 0;
        while level < self.tree.len() - 1 {
            let i_partner= Self::get_partner_index(i); 
            let proof_elem = ProofElement::new_from_index(self.tree[level][i_partner].clone(), i_partner);
            proof.push(proof_elem);
            level += 1;

            i /= 2;
        }
        Ok(proof)
    }
```

Plus in this way there isn't much repeated code. 

Please let me know what you think about this. I did run some tests and in both ways the output is the same and it is correct, but I think it looks clearer this way :)

Reply:
Yeah it is much more concise this way. Thanks! Very good suggestion.